### PR TITLE
add types to __traits(getLinkage,t)

### PR DESCRIPTION
**Pull request by WalterBright**
getLinkage needs to accept types, too

### CODE REVIEW

**Comment by WalterBright** [comment]
Blocking https://github.com/dlang/phobos/pull/5418

File: src/ddmd/traits.d [hunk 1]
```diff
@@ -917,27 +917,34 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
+        if (t)
         {
-            e.error("argument to `__traits(getLinkage, %s)` is not a declaration", o.toChars());
-            return new ErrorExp();
+            if (t.ty == Tfunction)
+                tf = cast(TypeFunction)t;
+            else if (t.ty == Tdelegate)
+                tf = cast(TypeFunction)t.nextOf();
+            else if (t.ty == Tpointer && t.nextOf().ty == Tfunction)
+                tf = cast(TypeFunction)t.nextOf();
```
Comment (ZombineDev) [inline src/ddmd/traits.d]: How about:
```D
        LINK link;
        auto o = (*e.args)[0];
        if (auto t = isType(o))
        {
            TypeFunction tf;
            if (t.ty == Tfunction)
                tf = cast(TypeFunction)t;
            else if (t.ty == Tdelegate || t.ty == Tpointer && t.nextOf().ty == Tfunction)
                tf = cast(TypeFunction)t.nextOf();
            assert(tf);
            link = tf.linkage;
        }
        else
        {
            auto s = getDsymbol(o);
            Declaration d;
            if (!s || (d = s.isDeclaration()) is null)
            {
                e.error("argument to `__traits(getLinkage, %s)` is not a declaration", o.toChars());
                return new ErrorExp();
            }
            link = d.linkage;
        }
        auto linkage = linkageToChars(link);
        auto se = new StringExp(e.loc, cast(char*)linkage);
```

File: src/ddmd/traits.d [hunk 2]
```diff
@@ -917,27 +917,34 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
+            link = tf.linkage;
+        else
         {
-            case LINK.d:        linkage = "D";           break;
-            case LINK.c:        linkage = "C";           break;
-            case LINK.cpp:      linkage = "C++";         break;
-            case LINK.windows:  linkage = "Windows";     break;
-            case LINK.pascal:   linkage = "Pascal";      break;
-            case LINK.objc:     linkage = "Objective-C"; break;
-            case LINK.system:   linkage = "System";      break;
```
Comment (ZombineDev) [inline src/ddmd/traits.d]: Nice simplification/refactoring

**Review by PetarKirov — APPROVED** [review]
A very nice enhancement over the ugly `mangledName!Func` hack, and I guess also saves a bit of compilation time.

File: src/ddmd/traits.d (hunk 1 above)
Comment (WalterBright) [inline src/ddmd/traits.d]: The `assert` will trip if someone passes a non-function type to the trait, so this change will not work correctly.

File: src/ddmd/traits.d [hunk 3]
```diff
@@ -917,27 +917,34 @@ extern (C++) Expression semanticTraits(TraitsExp e, Scope* sc)
...
-            default: assert(0);
+            auto s = getDsymbol(o);
+            Declaration d;
+            if (!s || (d = s.isDeclaration()) is null)
+            {
+                e.error("argument to `__traits(getLinkage, %s)` is not a declaration", o.toChars());
+                return new ErrorExp();
+            }
+            link = d.linkage;
```
Comment (andralex) [inline src/ddmd/traits.d]: cool job bro

**Review by andralex — APPROVED** [review]

**Pull request merged**